### PR TITLE
Provide a reliable way to see original resources in a directory

### LINKED ISSUE #25672: Listing directories in exported project returns only `*.import` files
**Godot version:**
Godot v3.0.6 Stable x64

**OS/device including version:**
Windows 8.1 

**Issue description:**

Shortly I tried to scan my `res://audio/supersounds` directory for every `.wav` file it had into array, and then play track by picking randomly with randi() % array.size(). But i've met problem where already exported project return empty array. So thinking about it might be problem with `.wav`, i've added some `.png` and `.ogg` into the folder and even removed file type check and it did some progress, it returned only `.import` type files. That's not what I wanted. In editor everything showed how expected.

Some guy in [godot questions topic](https://godotengine.org/qa/40340/gdscript-exported-project-return-empty-array) did reproduction and met the same problem. But he somehow got some `.png` and other types in output. Being surprised I also went and created empty project with same function to scan, but added custom `.png` into `res://` and tried to list the thing. This is what've got:

```
[Node.gd.remap, Node.gdc, Node.tscn, airSpike.png.import, default_env.tres, icon.png, icon.png.import, project.binary]
```

Apparently default files and main scene are somewhat defended against it!?

After... I went already here to search for answer I thought it might be [#25347](https://github.com/godotengine/godot/issues/25347), but i checked it's reproduction and it works fine, so file_exists() isn't the case.

Im at dead end now.

**Steps to reproduce:**
Create new scene, add root node, add following script to it:

```
func _ready():
	print(list_files_in_directory("res://"))

func list_files_in_directory(path):
	var files = []
	var dir = Directory.new()
	dir.open(path)
	dir.list_dir_begin()
	while true:
		var file = dir.get_next()
		if file == "":
			break
		elif not file.begins_with("."):
			files.append(file)
	dir.list_dir_end()
	return files
```
Export, launch. Executable debug console should print something similar to this:

```
[Node.gd.remap, Node.gdc, Node.tscn, airSpike.png.import, default_env.tres, icon.png, icon.png.import, project.binary]
```

Expected:

```
[Node.gd.remap, Node.gdc, Node.tscn, airSpike.png, airSpike.png.import, default_env.tres, icon.png, icon.png.import, project.binary]
```

**Minimal reproduction project:**
[exported.zip](https://github.com/godotengine/godot/files/2838169/exported.zip)
[project.zip](https://github.com/godotengine/godot/files/2838170/project.zip)

### PR DESCRIPTION
**Pull request by reduz**
When exporting a project, resources are often moved, converted or remapped (source import files are gone, text scenes are converted to binary, etc).

This new function is a convenience that allows to list a directory and obtain the actual original resource files.

Example

```GDScript
var resources = ResourceLoader.list_directory("res://images")
print(resources)
```
Result will be something like:
```
["image1.png","image2.png","image3.png"]
```
instead of
```
["image1.png.import","image2.png.import","image3.png.import"]
```

- *Production edit: This closes https://github.com/godotengine/godot/issues/25672.
See https://github.com/godotengine/godot/issues/66014.*

### CODE REVIEW

**Comment by AdriaandeJongh** [comment]
If you'll allow me to scrutinize the function name: as the function essentially lists files and not directories, perhaps an alternative name could be:
- `list_directory_files`
- `list_files_in_dir`
- `get_files_in_directory` (with `get_*` being perhaps more in line with other function names in ResourceLoader)
- `get_files`

**Comment by reduz** [comment]
It lists both files and directories, it just appends a "/" to the end of the directory,  I forgot to make note of this, will do so.

**Comment by JHDev2006** [comment]
Would this also mean that we wouldn't have to deal with .remap in exported games? Looks great! Looking forward to seeing this improvement!

**Comment by KoBeWi** [comment]
So I compared editor results with exported results and got this:
Editor `get_files_at()`
```
["Scene.tscn", "Script.gd", "icon.svg", "icon.svg.import", "textfile.txt"]
```
Editor `list_directory()`
```
["../", "./", "Folder/", "Scene.tscn", "Script.gd", "icon.svg"]
```
Exported `get_files_at()`
```
["Scene.tscn.remap", "Script.gd.remap", "Script.gdc", "icon.svg.import"]
```
Exported `list_directory()`
```
["Folder/", "Scene.tscn", "Script.gd", "Script.gdc", "icon.svg"]
```
Notes:
- Editor lists `.` and `..` for some reason. This should be removed.
- `list_directory()` does not include non-resource files. This should be stated in the documentation.

Other than that seems to work correctly.

File: core/io/resource_loader.cpp [hunk 1]
```diff
@@ -1354,6 +1356,53 @@ bool ResourceLoader::is_cleaning_tasks() {
...
+		return Vector<String>();
+	}
+
+	Error err = dir->list_dir_begin();
+	if (err != OK) {
+		return Vector<String>();
+	}
+
+	String d = dir->get_next();
+	while (d != String()) {
```
Comment (KoBeWi) [inline core/io/resource_loader.cpp:1372]:
```suggestion
	while (!d.is_empty()) {
```

File: core/io/resource_loader.cpp [hunk 2]
```diff
@@ -1354,6 +1356,53 @@ bool ResourceLoader::is_cleaning_tasks() {
...
+		bool recognized = false;
+		if (dir->current_is_dir()) {
+			d += "/";
+			recognized = true;
+		} else {
+			if (d.ends_with(".import") || d.ends_with(".remap")) {
+				int idx = d.rfind(".");
+				d = d.substr(0, idx);
```
Comment (KoBeWi) [inline core/io/resource_loader.cpp:1380]:
```suggestion
				d = d.substr(0, d.rfind("."));
```

File: core/io/resource_loader.cpp [hunk 3]
```diff
@@ -1354,6 +1356,53 @@ bool ResourceLoader::is_cleaning_tasks() {
...
+			}
+
+			String full_path = p_directory.path_join(d);
+			// Try all loaders and pick the first match for the type hint
```
Comment (KoBeWi) [inline core/io/resource_loader.cpp:1384]:
```suggestion
			const String full_path = p_directory.path_join(d);
			// Try all loaders and pick the first match for the type hint.
```

**Comment by Calinou** [comment]
Does this PR fully supersede https://github.com/godotengine/godot/pull/59334?

**Comment by KoBeWi** [comment]
Well my PR lists resource files together with non-resource files and excludes folders. But that's just alternative implementation, it's fully superseded.

**Comment by reduz** [comment]
Applied changes suggested.

File: core/io/resource_loader.cpp [hunk 4]
```diff
@@ -1439,6 +1441,60 @@ bool ResourceLoader::is_cleaning_tasks() {
...
+	Error err = dir->list_dir_begin();
+	if (err != OK) {
+		return Vector<String>();
+	}
+
+	String d = dir->get_next();
+	while (!d.is_empty()) {
+		bool recognized = false;
+		if (dir->current_is_dir()) {
+			if (d != "." && d != "..") {
```
Comment (KoBeWi) [inline core/io/resource_loader.cpp:1460]: It's weird that this is needed. DirAccess' `include_navigational` is false by default. The fact that it's different between editor and export means there is some bug.

**Review by KoBeWi — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

File: core/io/resource_loader.cpp (hunk 4 above)
Comment (KoBeWi) [reply core/io/resource_loader.cpp:1460]: I tried listing directory using `list_dir_begin()` in GDScript and it's correctly skipping navigational paths. Seems like `open()` in C++ works a bit differently (GDScript uses `_open()`), but not sure what causes the wrong behavior.

**Comment by nonchip** [comment]
please refer to comments in the PR this supersedes why it's not actually "reliable".

for example mine [there](https://github.com/godotengine/godot/pull/59334#issuecomment-1434253793) and [there](https://github.com/godotengine/godot/pull/59334#issuecomment-1436404105)

tldr: you're guessing that a certain filename means that it used to have a different one, but the mere existence of ExportPlugins makes that assumption questionable at best. like you said yourself in the original PR, there's no way of actually doing this "reliable" without creating a list during export. this PR is just copying a common gdscript hack/workaround to core instead.

**Comment by reduz** [comment]
@nonchip Yes, but that is a very corner case and generally external to the core engine, and even with export plugins remaps are still used, so it should work in almost every case. A "proper" solution is likely so complex that its not worth it.

**Review by akien-mga — APPROVED** [review]
